### PR TITLE
[MRG] DOC, Style: Use editor compatible code cells in gallery

### LINKED ISSUE #9563: DOC: Sphinx-gallery examples: replace bar of # characters with # %%
I suggest to replace the block of `#` characters with the more standard `# %%` in all sphinx-gallery examples.

I am talking about these blocks: https://github.com/mne-tools/mne-python/blob/308a0abf7c8b8b14c9afc6db2e3ed5b57787be31/tutorials/epochs/10_epochs_overview.py#L19-L23

The functionality for rendering the docs is equal, see: https://sphinx-gallery.github.io/stable/syntax.html#embed-rst-in-your-example-python-files

The advantage of using `# %%` is that many code editors recognize and parse `# %%` as "code blocks" so that you can execute an example similar to a notebook from your editor. This is good for playing around and debugging (using IPython). See the following quote from the docs (as linked above)

> The #%% and # %% syntax is consistent with the ‘code block’ (or ‘code cell’) separator syntax in Visual Studio Code Python extension, Visual Studio Python Tools, Jupytext, Pycharm Professional, Hydrogen plugin (for Atom) and Spyder. 

I switched to this style in many repos recently (e.g., mne-bids https://github.com/mne-tools/mne-bids/pull/830), and it's a good quality of live improvement for devs (to quickly try something) and users alike.

This could probably be done with a single search and replace operation. One thing that may be nice is to add a leading `# %%` **before** the imports, so that the imports become a code cell and can be executed. Currently they would not be parsed as a code cell, see the picture from above in a bigger view: https://github.com/mne-tools/mne-python/blob/308a0abf7c8b8b14c9afc6db2e3ed5b57787be31/tutorials/epochs/10_epochs_overview.py#L15-L23

--> code blocks would only start after the first block of `#` and that's undesirable.

What do you think?

### PR DESCRIPTION
**Pull request by sappelhoff**
closes #9563 

This is a global search and replace through the examples/ and tutorials/ directories for going from 79 `"#"` to `"# %% ####..."`.

I furthermore changed some licenses from "BSD style" to "BSD (3-clause)" (see 61d3f54f7b9c879fd506d660e80491c17d952e60) AND added licenses as follows to examples/tutorials that had no license:

```
# Authors: MNE-Developers
#
# License: BSD (3-clause)
```

(see a462d2e44d9839bdeaccc678823b0669bcff051f)

I then made use of the fact that each gallery file has a string `# License: BSD (3-clause)` before the imports to inject a `"# %% ####..."` right above the imports, so that editors will pick this up as a code cell as well.

Let's see how it looks. You can click on details to see the code I used for this.

If we later decide to use `# %%` instead of `# %% ######...`, that can be easily adjusted.

(PS: Going through the commits one by one is probably more helpful than looking at the whole diff in this case)

<details>

```python
"""Reformat sphinx-gallery.

Before running this I did a global search and replace from
###############################################################################
to the new # %% syntax

And I renamed several "BSD style" licenses to "BSD (3-clause)" for consistency
and clarity.

"""

# %%
# imports
from pathlib import Path
import glob
import io

# %%
# find all example and tutorial PYTHON files
p = Path("/home/stefanappelhoff/Desktop/bids/mne-python")

searchstr = str(p / "tutorials" / "**" / "*.py")
tutfiles = glob.glob(searchstr, recursive=True)
print(len(tutfiles))

searchstr = str(p / "examples" / "**" / "*.py")
exfiles = glob.glob(searchstr, recursive=True)
print(len(exfiles))

len(tutfiles) + len(exfiles)

# %%
# find files that are missing a license

match = "License: BSD (3-clause)"
fs_wo_lic = []
for fpath in tutfiles + exfiles:
    with open(fpath, "r") as fin:
        lines = fin.readlines()

    flag = True
    for line in lines:
        if match in line:
            flag = False
            break

    if flag:
        fs_wo_lic.append(fpath)

fs_wo_lic

# %%
# make sure each example / tutorial has a copyright line at the top

inject = """
# Authors: MNE-Developers
#
# License: BSD (3-clause)
"""
inject_lines = io.StringIO(inject).readlines()


for fpath in fs_wo_lic:
    with open(fpath, "r") as fin:
        lines = fin.readlines()

    new_lines = []
    got_quotes = 0
    injection_done = False
    for line in lines:

        # We'll print this line back later
        new_lines += line

        # Now check injection
        if '"""' in line:
            got_quotes += 1

        if got_quotes == 2 and not injection_done:
            new_lines += inject_lines
            injection_done = True

    with open(fpath, "w") as fout:
        fout.writelines(new_lines)

# %%
# Now make the imports their own cell

inject_part = "# %%"
inject = f"""
{inject_part} ##########################################################################
"""
inject_lines = io.StringIO(inject).readlines()

for fpath in tutfiles + exfiles:
    with open(fpath, "r") as fin:
        lines = fin.readlines()

    new_lines = []
    injection_done = False
    for line in lines:

        # We'll print this line back later
        new_lines += line

        # Now check injection
        if "# License: BSD (3-clause)" in line and not injection_done:
            new_lines += inject_lines
            injection_done = True

    with open(fpath, "w") as fout:
        fout.writelines(new_lines)

# %%

```
</details>

### CODE REVIEW

**Comment by drammock** [comment]
let's stick with `# %%` after all.  It will be easier to type when writing new tutorials, and I'm OK with how it ends up looking in the editor (given that many editors do add an extra line or something that makes it more "landmarky").

**Comment by sappelhoff** [comment]
Okay thanks @drammock, I'll do that in the next commit. Before doing that, could you inspect the rendered docs and let me know if I should change anything?

- Before: https://mne.tools/dev/auto_examples/connectivity/mixed_source_space_connectivity.html
- After: https://30859-1301584-gh.circle-artifacts.com/0/dev/auto_examples/connectivity/mixed_source_space_connectivity.html#sphx-glr-auto-examples-connectivity-mixed-source-space-connectivity-py

I did only a few spot checks:

- nothing should have changed in the rendering
- EXCEPT that the "Authors and license" snippet should be its own cell now, which I think is okay. WDYT?

Before:

image


After:

image

**Comment by drammock** [comment]
I took a look, and honestly I'm not crazy about the addition of
```
# Authors: MNE-Developers
# License: BSD (3-clause)
```
(e.g., here: https://30859-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/intro/10_overview.html compared to current main https://mne.tools/stable/auto_tutorials/intro/10_overview.html).

My reasons are:
1. authorship is trackable by `git blame`, I never really saw the point of putting it in the examples/tutorials to begin with
2. saying the author is "MNE Developers" is uninformative
3. the license is already specified in the repo, seems unnecessary to re-state it on every example/tutorial

If anything, I'd lean toward *removing* the author/license block on tutorials/examples that have it, rather than adding a (redundant?) author/license block to ones that didn't have it yet.

@sappelhoff in `requirements_doc.txt` can you also change `sphinx` to  `sphinx != 4.1.0` please?  There's a bug in 4.1.0 that makes our logo disappear (see the circleci rendering, it just says "logo" there); a fix is merged already but 4.1.1 hasn't been released yet.

**Comment by sappelhoff** [comment]
I think these are valid reasons and I'd agree with all of them in principle.

One valid point for having the author/copyright notices is "recognition" (e.g., for new contributors), and maybe a point to a lesser degree is "easier accountability" (not everybody has the knowledge to search through git blame).

The main reason why I added them here is consistency: This was only missing in a minority of cases, and in an even smaller minority it said "BSD style" instead of BSD 3-clause.

Having that said, I would be up to removing all of these notices (consistently). I just think that it would need a bigger discussion than adding the few notices with "MNE Devs" as copyright holder, just because removing somebody's name is maybe more of an affront. I don't know, interested in your view here. Could also be done in a separate PR.

I'll add the sphinx change in the next push.

**Comment by drammock** [comment]
After some off-GitHub discussion, the emergent decision seems to be:

1. change `BSD-style` to `BSD 3-clause`
2. don't add `# authors: MNE Developers` anywhere
3. leave existing authors alone in this PR; we'll do a separate PR to remove them (but tag everyone who would be affected, and be open to not merging it if there's general disapproval)

To this I will add my personal aesthetic opinion that it looks odd to me when there's a code cell in the rendered tutorial that only contains author/license comments.  If there's an *easy* way to get the existing author/license comments into the imports code cell (presumably by putting the `# %%` a couple lines higher), I would advocate for that.

**Comment by rob-luke** [comment]
> leave existing authors alone in this PR; we'll do a separate PR to remove them 

I'm -1 on this. While git blame is possible, the number of users who know about it and can use it (even with GitHub button) are the minority. Writing the tutorials is extremely time consuming. I feel that having your name at the top of these documents is good for recognition amongst ones peers, and this would be lost if they are removed.

For those people who's names are on the papers this probably makes no difference. But for the smaller contributors this is probably a big deal. Just as we added new authors to the change log, I think including names at top of examples is good encouragement for new submitters.

If anything I would vote to also add all reviewers of a tutorial to increase contributor exposure, not reduce it.

**Comment by rob-luke** [comment]
I think in this situation it's good to put on the non-developer hat. It's likely only coders read the change logs. Whereas your PIs and general scientists probably read the tutorial/examples. So for widest possible recognition of contributions the tutorials would be most effective

**Comment by sappelhoff** [comment]
Thanks for the summary Dan, and thanks for chiming in from a different perspective Rob.

For now I'll:

1. leave my change from BAS-style --> BSD 3-clause
2. remove my change to add generic "MNE Developers" copyright notices

Lastly, regarding your comment:

> To this I will add my personal aesthetic opinion that it looks odd to me when there's a code cell in the rendered tutorial that only contains author/license comments. If there's an easy way to get the existing author/license comments into the imports code cell (presumably by putting the # %% a couple lines higher), I would advocate for that.

I think it's actually nicer to have the imports cell without the authors, so that If I want to actually copy the imports, I don't have to remove the authors from my own script into which I want to copy the imports. But I agree with your point that having the authors as their own cell looks weird. Perhaps I can try to integrate them into the initial docstring (bounded by the triple quotation marks)

**Comment by sappelhoff** [comment]
> But I agree with your point that having the authors as their own cell looks weird. Perhaps I can try to integrate them into the initial docstring (bounded by the triple quotation marks)

I ended up **not** changing this from my previous commits: Authors/License are still their own code block before the imports codeblock.

1. I don't find that nice, but **nicer** than having a comment about authors in the imports block
2. We might (or not, we'll see how the discussion goes) remove the author/license blocks anyhow.

This is ready for review / merge from my side.

**Review by agramfort — APPROVED** [review]
LGTM assuming it's just replacements from:
###############################################################################
to
# %%

**Comment by sappelhoff** [comment]
Yes, it's:

- replace `###############` to `# %%`
- block sphinx 4.1.0
- change BSD-style --> BSD 3-clause
- Add a `#%%` **before** the imports

all of this only in tutorials/ and examples/

File: tutorials/simulation/80_dics.py [hunk 1]
```diff
@@ -15,7 +15,9 @@
 #
 # License: BSD (3-clause)
 
-###############################################################################
+# %%
+
+# %%
```
Comment (drammock) [inline tutorials/simulation/80_dics.py:20]: again

**Review by drammock — APPROVED** [review]
other than comments LGTM.  Reviewing this made me realize we should a "how to write examples/tutorials" section to the coding conventions section of our contributing guide. I'll open a separate issue about that.

File: examples/visualization/publication_figure.py [hunk 2]
```diff
@@ -15,7 +15,9 @@
 #
 # License: BSD (3-clause)
 
-###############################################################################
+# %%
+
+# %%
```
[the remark this answers is not in the document]
Comment (sappelhoff) [reply examples/visualization/publication_figure.py:20]: thanks for catching these. I think I inadvertently added those via code when import sections **already** had a code cell.

**Comment by hoechenberger** [comment]
Sorry I'm a bit late, just wanted to share that to SPDX identifier for the license is `BSD-3-Clause` (mind the dashes), see
https://spdx.org/licenses/

**Comment by sappelhoff** [comment]
I think it's a good idea to use that identifier, so I made some global search+replace calls.

There remain some BSD licenses in the code base that I did not remain because it was unclear:

- BSD 2-clause --> BSD-2-Clause?
- BSD Style --> BSD-3-Clause?
- BSD (or similar) --> ??
- BSD form
- BSD license

**Comment by hoechenberger** [comment]
> * BSD 2-clause --> BSD-2-Clause?

Yes

> * BSD Style --> BSD-3-Clause?
> * BSD (or similar) --> ??
> * BSD form
> * BSD license

I think these are all under-specified and we'd have to guess :\

**Comment by sappelhoff** [comment]
okay, done, except for one case of "BSD 2-clause" which was a copy paste from somewhere, and that's probably something we don't want to touch.

As for the rest: I vote leaving them alone for now.

**Comment by hoechenberger** [comment]
> As for the rest: I vote leaving them alone for now.

+1

**Review by drammock — APPROVED** [review]

**Comment by drammock** [comment]
+1 for merge when CIs green

**Comment by hoechenberger** [comment]
Thanks @sappelhoff!

**Pull request merged**